clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -1250,26 +1246,52 @@ protected void updateOffsets(
   @Override
   protected void updateOffsetMetadataInOffsetRecord(
       PartitionConsumptionState partitionConsumptionState,
-      OffsetRecord offsetRecord,
-      ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecordWrapper,
-      LeaderProducedRecordContext leaderProducedRecordContext,
-      String kafkaUrl) {
-    updateOffsets(
+      ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecord,
+      String upstreamKafkaUrl,
+      boolean isShutdown) {
+    updateOffsetsFromPartitionConsumptionState(
         partitionConsumptionState,
-        consumerRecordWrapper,
-        leaderProducedRecordContext,
-        offsetRecord::setCheckpointLocalVersionTopicOffset,
-        (sourceKafkaUrl, upstreamTopicName, upstreamTopicOffset) -> {
-          if (Version.isRealTimeTopic(upstreamTopicName)) {
-            offsetRecord.setLeaderUpstreamOffset(sourceKafkaUrl, upstreamTopicOffset);
-          } else {
-            offsetRecord.setCheckpointUpstreamVersionTopicOffset(upstreamTopicOffset);
-          }
-        },
-        (sourceKafkaUrl, upstreamTopicName) -> Version.isRealTimeTopic(upstreamTopicName)
-            ? offsetRecord.getUpstreamOffset(sourceKafkaUrl)
-            : offsetRecord.getCheckpointUpstreamVersionTopicOffset(),
-        () -> OffsetRecord.NON_AA_REPLICATION_UPSTREAM_OFFSET_MAP_KEY);
+        () -> OffsetRecord.NON_AA_REPLICATION_UPSTREAM_OFFSET_MAP_KEY,
+        shouldUpdateUpstreamOffset(consumerRecord),
+        isShutdown);
+  }
+
+  /**
+   * Sync the metadata about offset from {@link PartitionConsumptionState} to {@link OffsetRecord}.
+   * It's agnostic to the record as {@link PartitionConsumptionState} should have handled it already.
+   * @param partitionConsumptionState
+   * @param sourceKafkaUrl, The computed upstream Kafka URL for the record.
+   * @param shouldUpstreamOffset, the flag to update the upstream offset or not. If the record contains invalid upstream
+   *                              offset, then upstream kafka url doesn't need to be computed
+   * @param isShutdown, indicating if the method is called for shut down event. If so, offsetRecord will reset its upstream
+   *                   offset map.
```

Comment:
The doc says that the method is agnostic to record, but `sourceKafkaUrl` and `shouldUpstreamOffset` params still come from record. Could we completely use PCS to update OffsetRecord, for both shutdown or non-shutdown cases?

Reply:
I really wish we can completely use PCS to update OffsetRecord but I am uncertain about this due to the existing logic of updating upstream offsets.

```
        if (Version.isRealTimeTopic(upstreamTopicName)) {
          offsetRecord.setLeaderUpstreamOffset(sourceKafkaUrl.get(), newUpstreamOffset);
        } else {
          offsetRecord.setCheckpointUpstreamVersionTopicOffset(newUpstreamOffset);
        }
```
During the regular consumption, if that's a real time topic, then it has to specify the upstream kafka url. In order to compute the upstream kafka url, we need `shouldUpstreamOffset` to validate it, which requires a consumer record.

For L/F SIT, that's not a problem because the sourceKafkaUrl is always `NON_AA_REPLICATION_UPSTREAM_OFFSET_MAP_KEY` but for the AA SIT, it requres a bit [different logic](https://github.com/linkedin/venice/blob/4772a3643df9f0fdbae59c46de2ba7ab29ca3f07/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java#L954).

So, I think we still need to rely on it unless there's a better way to update the upstream offset.

Reply:
During updateLatestInMemoryProcessedOffset, latestProcessedUpstreamRTOffsetMap is already updated with kafka url and offset, which means latestProcessedUpstreamRTOffsetMap should be up-to-date. Could we use latestProcessedUpstreamRTOffsetMap to update OffsetRecord's map?

Reply:
Ack. I think we can do it!